### PR TITLE
feat: add studio write access validation

### PR DESCRIPTION
**Pull request by germanolleunlp**
# Add studio write access validation

### Ticket: https://jira.2u.com/browse/ACADEMIC-16359

**Merge checklist:**
Check off if complete *or* not applicable:
- [X] Version bumped
- [X] Changelog record added
- [X] Documentation updated (not only docstrings)
- [X] Fixup commits are squashed away
- [X] Unit tests added/updated
- [X] Manual testing instructions provided
- [X] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

### CODE REVIEW

File: ai_aside/platform_imports.py [hunk 1]
```diff
@@ -26,3 +26,15 @@ def get_block(usage_key):
     # pylint: disable=import-error, import-outside-toplevel
     from xmodule.modulestore.django import modulestore
     return modulestore().get_item(usage_key)
+
+
+def can_change_summaries_settings(user, course_key) -> bool:
+    """
+    Check if the user can change the summaries settings.
```
Comment (ashultz0) [inline ai_aside/platform_imports.py:33]: worth putting "by checking for studio write access" in this comment

File: ai_aside/api/__init__.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        super().__init__(data=reply, status=_status, content_type=content_type)
+
+
+class AiAsideAPIView(APIView):
+    """
+    Overrides APIView to handle ai_aside exceptions
+    """
+    def handle_exception(self, exc):
+        """
+        Converts proctoring exceptions into standard restframework responses
```
Comment (ashultz0) [inline ai_aside/api/__init__.py:40]: copy error, should be aside exceptions

File: tests/api/test_views.py [hunk 3]
```diff
@@ -20,10 +20,25 @@
     'block-v1:edX+DemoX+Demo_Course+type@vertical+block@vertical_321ac313f2de',
 ]
 
+has_studio_write_access = True
```
Comment (ashultz0) [inline tests/api/test_views.py:23]: you only test the true case here

File: tests/api/test_views.py [hunk 4]
```diff
@@ -20,10 +20,25 @@
...
+def fake_has_studio_write_access(user, course_key):  # pylint: disable=unused-argument
+    return (has_studio_write_access, 'unused', 'unused')
+
 
 @ddt.ddt
 class TestApiViews(APITestCase):
     """API Endpoint View tests"""
+    def setUp(self):
+        auth_mock = MagicMock()
+        auth_mock.has_studio_write_access = fake_has_studio_write_access
```
Comment (ashultz0) [inline tests/api/test_views.py:35]: I think there are mock methods where you can say auth_mock.has_studio_write_access is a method that returns true, then you don't need the fake and it would be easy to test the negative case too

File: ai_aside/config_api/views.py [hunk 5]
```diff
@@ -34,47 +33,36 @@
...
+        course_key = validate_course_key(course_id)
+        enabled = is_summary_config_enabled(course_key)
+        return APIResponse(success=True, data={'enabled': enabled})
 
 
-class CourseEnabledAPIView(APIView):
+class CourseEnabledAPIView(AiAsideAPIView):
     """Handlers for course level settings"""
 
+    permission_classes = (HasStudioWriteAccess,)
```
Comment (ashultz0) [inline ai_aside/config_api/views.py:56]: I think you also need to add the jwt auth decorator or the endpoint won't have decoded the user. But I might be wrong.

have you tested this locally yet?

File: ai_aside/api/view_utils.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    Validate and parse a unit_key string, if supported.
+
+    Args:
+        unit_key_string (str): string unit key to validate
+
+    Returns:
+        UsageKey: validated unit key
+
+    Raises:
+        ValidationError: DRF Validation error in case the unit key is invalid
```
Comment (ashultz0) [inline ai_aside/api/view_utils.py:43]: this comment isn't the way it works here, wrong error type

also a very wordy style of docstring, not my favorite but fine if you want it

File: ai_aside/api/view_utils.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+"""
+Utilities related to API views
```
Comment (ashultz0) [inline ai_aside/api/view_utils.py:2]: this file is actually just validators.py

unless you take the stuff in init and put it in here which might be nice, things often get missed in init.py

File: ai_aside/api/__init__.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
+"""
```
Comment (ashultz0) [inline ai_aside/api/__init__.py:1]: we don't have a generic api directory right now and this stuff is all specific to the config api, I think it goes down at that level rather than generalizing to the overall API we don't have

File: ai_aside/api/permissions.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+""" Permissions for ai-aside API"""
```
Comment (ashultz0) [inline ai_aside/api/permissions.py:1]: this one is especially config-api specific

**Review by ashultz0 — CHANGES_REQUESTED** [review]
needs a changelog entry and update of version number, probably to 3.5.0

thoughts about organization you can take or leave

also it might need the decorator that actually requires and decodes a user on the view calls

File: ai_aside/platform_imports.py (hunk 1 above)
Comment (germanolleunlp) [reply ai_aside/platform_imports.py:33]: ✅

File: ai_aside/api/__init__.py (hunk 2 above)
Comment (germanolleunlp) [reply ai_aside/api/__init__.py:40]: ✅

File: ai_aside/config_api/views.py (hunk 5 above)
Comment (germanolleunlp) [reply ai_aside/config_api/views.py:54]: I tested it locally, and it was working without the `Jwt` decorator

Comment (germanolleunlp) [reply ai_aside/config_api/views.py:54]: and it has the user 👍🏽

File: ai_aside/api/view_utils.py (hunk 6 above)
Comment (germanolleunlp) [reply ai_aside/api/view_utils.py:44]: ✅

File: ai_aside/api/view_utils.py (hunk 7 above)
Comment (germanolleunlp) [reply ai_aside/api/view_utils.py:2]: ✅

File: ai_aside/api/__init__.py (hunk 8 above)
Comment (germanolleunlp) [reply ai_aside/api/__init__.py:1]: removed and moved it to `config_api`

File: ai_aside/api/permissions.py (hunk 9 above)
Comment (germanolleunlp) [reply ai_aside/api/permissions.py:1]: ✅

File: tests/api/test_views.py (hunk 3 above)
Comment (germanolleunlp) [reply tests/api/test_views.py]: ✅

File: tests/api/test_views.py (hunk 4 above)
Comment (germanolleunlp) [reply tests/api/test_views.py]: ✅ replaced with @patch

**Comment by germanolleunlp** [comment]
> needs a changelog entry and update of version number, probably to 3.5.0
> 
> thoughts about organization you can take or leave
> 
> also it might need the decorator that actually requires and decodes a user on the view calls

💹  done

File: tests/api/test_views.py [hunk 10]
```diff
@@ -319,3 +331,85 @@ def test_course_enabled_setter_enable_valid_and_reset(self):
...
+
+
+class TestApiViewsWithoutPermissions(AIAsideAPITestCase):
+    """API Endpoint View tests without permissions"""
+    def setUp(self):
+        super().setUp()
+        can_change_summaries_settings.return_value = False
+        self.access_mock = patch('ai_aside.platform_imports.can_change_summaries_settings',
+                                 can_change_summaries_settings)
+        self.access_mock.start()
```
Comment (ashultz0) [inline tests/api/test_views.py:343]: this is a weird python resource leak technicality

in super.setup it makes self.accessmock and starts it, let's call that mock1
then here you reassign self.accessmock to a new mock and start it, that's mock2
there is no longer any reference available to mock1, self.accessmock is now mock2

in teardown we tear down mock2. mock1 is never torn down.

I don't think that mocks really leak anything significant especially if you never start them so this is just something to notice. Fixing it would just be not calling super setup or delegating mock creation into a function which reads self.can_change_summaries or a million other methods.

Comment (germanolleunlp) [reply tests/api/test_views.py:343]: Thanks!
